### ISSUE TITLE
InvalidCastException, object of type 'LiteDB.EmptyPage' to type 'LiteDB.HeaderPage', on collection Insert()

### ISSUE THREAD

**Issue by Grul-** [mbdavid/LiteDB#811]
Hello,
I get the following error when I'm trying to add a new object to an existing collection:

```
System.InvalidCastException: Unable to cast object of type 'LiteDB.EmptyPage' to type 'LiteDB.HeaderPage'.
   in LiteDB.PageService.GetPage[T](UInt32 pageID)
   in LiteDB.TransactionService.Commit()
   in LiteDB.LiteEngine.Commit()
   in LiteDB.LiteEngine.Transaction[T](String collection, Boolean addIfNotExists, Func`2 action)
   in LiteDB.LiteEngine.Insert(String collection, BsonDocument doc)
   in LiteDB.LiteCollection`1.Insert(T document)

```
I'm not able at the moment to provide a method to replicate this exception, because I'm not able to identify a fixed event that triggers the error. By now the only consideration I can do is that I noticed this exception occurs when the collection exceeds a number of elements, but I'm not sure about this.

Do you have any idea about this?

Actually I'm using v3.1.4 but I used to encounter this error on previous 3.x versions too.

**Comment by mbdavid** [comment]
Hi @Grul-, this it's a v3 bug on concurrency environment. Please, consider upgrade to v4 where this is fixed... or, if you want keep in v3 you will need wrap database instance with lockers..

**Comment by Grul-** [comment]
Hi,
Upgrading to v4 solved that Exception.
But now I have another issue to solve, the objects I previously stored (with v3) don't have an Id property, but I used a property called Guid instead.
Now if I add a new Id property to my class I get an `InvalidCastException` during deserialization, on the other hand, if I don't add the Id property, I get a `LiteDB.LiteException: 'Invalid BSON data type 'Null' on field '_id'.'`. I've also tried to add the `[BsonId]` attribute to my Guid property but the cast is still not valid.

Is there a way to keep the existing property and to use it in the BsonDocument as Id?

**Comment by mbdavid** [comment]
Hi,

If (in v2) your class has no Id and you set Guid id in method, LiteDB just set in document (after deserialize your class to BsonDocumento) `_id = yourGuid`. You can check this on shell tool. In shell/json, Guid must me show as `_id: { $guid: "00000000-0000-0000-0000-000000000000" }`. Checks if is not a string value.

If you want add Id now, you need Id with Guid type. You can still using Insert without an Id, as here: https://github.com/mbdavid/LiteDB/blob/master/LiteDB/Database/Collections/Insert.cs#L33

**Comment by Grul-** [comment]
I guess I found the problem. I used the shell tool to open the existing database file and (here it is!) the deserialized objects have their `_id` defined as `"_id":{"$oid":"5a0571ac5083f01ee8b8161a"}`. 
So I guess the InvalidCastException is due to the `_id` type being `$oid` and not `$guid`. So I add the Id property, as I tried before, but I changed its type from `System.Guid` to `LiteDB.ObjectId`, now everything works fine.

Thank you for your answers!
